src/Repository/RedirectRepository.php

```diff
@@ -7,45 +7,98 @@
 use Doctrine\ORM\NonUniqueResultException;
+use Exception;
 use Setono\SyliusRedirectPlugin\Model\RedirectInterface;
 use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Channel\Model\ChannelInterface;
 
 class RedirectRepository extends EntityRepository implements RedirectRepositoryInterface
 {
+    /**
+     * @throws Exception
+     */
+    public function removeNotAccessed(int $threshold): void
+    {
+        if ($threshold <= 0) {
+            return;
+        }
+
+        $dateTimeThreshold = (new DateTime())->sub(new DateInterval('P' . $threshold . 'D'));
+
+        $this->createQueryBuilder('r')
+            ->delete()
+            ->andWhere('r.lastAccessed is not null')
+            ->andWhere('r.lastAccessed <= :threshold')
+            ->setParameter('threshold', $dateTimeThreshold)
+            ->getQuery()
+            ->execute()
+        ;
+    }
+
+    public function findEnabledBySourceAndChannel(string $source, ChannelInterface $channel, bool $only404 = false): ?RedirectInterface
+    {
+        $qb = $this->createQueryBuilder('o')
+            ->select('o, c')
+            ->andWhere('o.source = :source')
+            ->andWhere('o.enabled = true')
+            ->setParameter('source', $source)
+            ->leftJoin('o.channels', 'c')
+        ;
+
+        if ($only404) {
+            $qb->andWhere('o.only404 = true');
+        }
+
+        /** @var RedirectInterface[] $redirects */
+        $redirects = $qb->getQuery()->getResult();
+
+        $preferredRedirect = null;
+
+        foreach ($redirects as $redirect) {
+            if (null === $preferredRedirect) {
+                $preferredRedirect = $redirect;
+
+                continue;
+            }
+
+            if ($redirect->getChannels()->count() > $preferredRedirect->getChannels()->count()) {
+                $preferredRedirect = $redirect;
+            }
+        }
+
+        return $preferredRedirect;
+    }
+
     /**
      * @throws NonUniqueResultException
      */
-    public function findEnabledBySource(string $source, bool $only404 = false): ?RedirectInterface
+    public function findEnabledBySource(string $source, bool $only404 = false, bool $fetchJoinChannels = false): ?RedirectInterface
     {
         $qb = $this->createQueryBuilder('r')
             ->andWhere('r.source = :source')
             ->andWhere('r.enabled = true')
+            ->setMaxResults(1)
             ->setParameter('source', $source);
 
         if ($only404) {
             $qb->andWhere('r.only404 = true');
         }
 
+        if ($fetchJoinChannels) {
+            $qb->select('r, c')
+                ->leftJoin('r.channels', 'c')
+            ;
+        }
+
         return $qb->getQuery()
             ->getOneOrNullResult();
     }
 
-    public function removeNotAccessed(int $threshold): void
+    public function searchNextRedirectByChannel(RedirectInterface $redirect, ChannelInterface $channel, bool $only404 = false): ?RedirectInterface
```

Comment:
Looks like we can omit this function. Or it required for BC?

Reply:
Yes, we can't remove any methods :(